### PR TITLE
bug(batch-action, product) Batch action query optimization

### PR DESCRIPTION
**Pull request by wiewiurdp**
# Description

Batch action query optimization

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Unit Tests
- [x] Behat Test

# Checklist:
- [x] I have read the contribution requirements and fulfill them.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: module/batch-action/src/Infrastructure/Persistence/Query/DbalBatchActionQuery.php [hunk 1]
```diff
@@ -54,7 +54,7 @@ public function getInformation(BatchActionId $id, Language $language): BatchActi
             $record['all'],
             $record['processed'],
             new \DateTime($record['created_at']),
-            $record['all'] === $record['processed'] ? new \DateTime($record['last_processed_at']) : null
+            $record['all'] === $record['processed'] ? new \DateTime($record['last_processed_at'] ?? $record['created_at']) : null
```
Comment (piotrkreft) [inline module/batch-action/src/Infrastructure/Persistence/Query/DbalBatchActionQuery.php:57]: This seems redundant and not related

File: module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        $queryBuilder = $this->getQueryBuilder();
+        if ($filter) {
+            $filterQuery = $filter->getQuery();
+            if ($filterQuery) {
+                if ($filter->getIds()) {
+                    if ($filter->getIds()->isIncluded()) {
+                        //query and include id
+                        $filteredQueryBuilder = $this->getFilteredQueryBuilder($filterQuery);
+
+                        return $filteredQueryBuilder->andWhere($filteredQueryBuilder->expr()->in('id', ':productsIds'))
```
Comment (piotrkreft) [inline module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php:54]: Shouldn't this be orWhere? Wouldn't it delimit result to only those ids?

**Review by piotrkreft — CHANGES_REQUESTED** [review]

**Review by piotrkreft — DISMISSED** [review]

File: module/product/features/simple-product-batch-action-delete.feature [hunk 3]
```diff
@@ -15,7 +15,7 @@ Feature: batch action product deletion
     Then the response status code should be 201
     And store response param "id" as "product_template"
 
-  Scenario: Create product
+  Scenario: Create product 1
```
Comment (wfajczyk) [inline module/product/features/simple-product-batch-action-delete.feature:18]:
```suggestion
Scenario Outline: Create product <type> 
```

File: module/product/features/simple-product-batch-action-delete.feature [hunk 4]
```diff
@@ -45,12 +70,39 @@ Feature: batch action product deletion
...
+    Then the response status code should be 201
+    And store response param "id" as "batch_action_1_id"
+
   Scenario: Get not exists batch action
     And I send a "GET" request to "/api/v1/en_GB/batch-action/@batch_action_1_id@"
     Then the response status code should be 200
 
-  Scenario: Delete product alredy deleted
-    When I send a DELETE request to "/api/v1/en_GB/products/@product_id@"
+  Scenario: Delete product 1 already deleted
```
Comment (wfajczyk) [inline module/product/features/simple-product-batch-action-delete.feature:96]:
```suggestion
  Scenario Outline: Delete product <name> already deleted
```

File: module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            }
+        }
+
+        //no filter or empty, get all items
+        return $queryBuilder;
+    }
+
+    private function getFilteredQueryBuilder(string $filter): QueryBuilder
+    {
+        $language = new Language('en_GB');
```
Comment (wfajczyk) [inline module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php:89]:
```suggestion
        $language = $this->languageQuery->getRootLanguage();
```
Where 
`$languageQuery` instanceof `Ergonode\Core\Domain\Query\LanguageQueryInterface;`

File: module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                        $filteredQueryBuilder = $this->getFilteredQueryBuilder($filterQuery);
+
+                        return $filteredQueryBuilder->orWhere($filteredQueryBuilder->expr()->in('id', ':productsIds'))
+                            ->setParameter(':productsIds', $filter->getIds()->getList(), Connection::PARAM_STR_ARRAY);
+                    }
+
+                    //query and exclude id
+                    $filteredQueryBuilder = $this->getFilteredQueryBuilder($filterQuery);
+
+                    return $filteredQueryBuilder->orWhere($filteredQueryBuilder->expr()->notIn('id', ':productsIds'))
```
Comment (wfajczyk) [inline module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php:62]:
```suggestion
                    return $filteredQueryBuilder->andWhere($filteredQueryBuilder->expr()->notIn('id', ':productsIds'))
```
Example:
- [A,B,C] -> or not [B,D] => [A,B,C]  is not expected,
- [A,B,C] -> and not [B,D] => [A, C]  is expected,

**Review by wfajczyk — CHANGES_REQUESTED** [review]

File: module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                }
+
+                //only query
+                return $this->getFilteredQueryBuilder($filter->getQuery());
+            }
+
+            if ($filter->getIds()) {
+                if ($filter->getIds()->isIncluded()) {
+                    //only include ids
+                    return $queryBuilder->andWhere($queryBuilder->expr()->in('id', ':productsIds'))
```
Comment (piotrkreft) [inline module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php:73]:
```suggestion
                    return $queryBuilder->orWhere($queryBuilder->expr()->in('id', ':productsIds'))
```

Comment (wfajczyk) [reply module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php:73]: IN this place is only id `andWhere` is ok.

Comment (piotrkreft) [reply module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php:73]: Yeah - actually it should be just `where` for clarity I guess

**Review by wfajczyk — APPROVED** [review]

File: module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    ) {
+        $this->productQueryBuilderFactory = $productQueryBuilderFactory;
+        $this->gridBuilder = $gridBuilder;
+        $this->productIdsProvider = $productIdsProvider;
+        $this->connection = $connection;
+        $this->languageQuery = $languageQuery;
+    }
+
+
+    public function provide(?BatchActionFilter $filter): QueryBuilder
```
Comment (rprzedzik) [inline module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php:48]: Make it required and check that filter exist (or even $filter->getQuery()) move service that use this provider.
Also IMHO this is more builder then provider .

**Review by rprzedzik — CHANGES_REQUESTED** [review]

File: module/product/src/Infrastructure/Provider/FilteredQueryBuilder.php [hunk 9]
```diff
@@ -45,7 +45,7 @@ public function __construct(
     }
 
 
-    public function provide(?BatchActionFilter $filter): QueryBuilder
+    public function build(BatchActionFilter $filter): QueryBuilder
     {
         $queryBuilder = $this->getQueryBuilder();
         if ($filter) {
```
Comment (rprzedzik) [inline module/product/src/Infrastructure/Provider/FilteredQueryBuilder.php:51]: This "if" is not required anymore

**Review by rprzedzik — CHANGES_REQUESTED** [review]

**Review by rprzedzik — DISMISSED** [review]

File: module/product/src/Infrastructure/Provider/FilteredQueryBuilder.php [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+        if ($filter->getIds()) {
+            if ($filter->getIds()->isIncluded()) {
+                //only include ids
+                return $queryBuilder->where($queryBuilder->expr()->in('id', ':productsIds'))
+                    ->setParameter(':productsIds', $filter->getIds()->getList(), Connection::PARAM_STR_ARRAY);
+            }
+
+            //only exclude id in all system
+            return $queryBuilder->andWhere($queryBuilder->expr()->notIn('id', ':productsIds'))
```
Comment (piotrkreft) [inline module/product/src/Infrastructure/Provider/FilteredQueryBuilder.php:81]:
```suggestion
            return $queryBuilder->where($queryBuilder->expr()->notIn('id', ':productsIds'))
```

**Review by piotrkreft — CHANGES_REQUESTED** [review]

**Review by rprzedzik — APPROVED** [review]

**Review by piotrkreft — APPROVED** [review]

**Review by wfajczyk — APPROVED** [review]

**Pull request merged**